### ISSUE TITLE
GCC 8.2.1 strncpy build error

### ISSUE THREAD

**Issue by mablem8** [ARM-software/armnn#140]
Using aarch64-linux-gnu-g++ (GNU Toolchain for the A-profile Architecture 8.2-2018.11 (arm-rel-8.26)) 8.2.1 20180802 from the [Arm GNU-A](https://developer.arm.com/open-source/gnu-toolchain/gnu-a/downloads) downloads page causes errors when compiling Arm NN Release 18.11 (tag v18.11).

One error, catching polymorphic type by value (-Werror=catch-value=) is reported [here](https://github.com/ARM-software/armnn/issues/123) and fixed with [this](https://github.com/ARM-software/armnn/pull/124) pull request.

The other error, strncpy specified bound depends on the length of the source argument (-Werror=stringop-overflow=) has not been reported here as an issue. One fix is to change strncpy to memcpy in src/armnn/LayerSupport.cpp.

**Comment by MatthewARM** [comment]
Hi @mablem8 this issue has been noticed before on the armnn-dev mailing list (https://lists.linaro.org/mailman/listinfo/armnn-dev). Nicolas on that list also suggested using a memcpy-based workaround, but I'm not sure because that just silences the warning, it doesn't solve the potential problem that g++ is indicating.

I was thinking that we shouldn't do the "std::min(maxLength, strlen(fullString))" bit, just pass the length of the destination buffer as the length argument to strncpy. That will copy extra zeroes, which is fine in this piece of error handling code.

What do you think?

Also, sorry to be a pain but please could submit patches as Gerrit reviews at https://review.mlplatform.org/#/admin/projects/ml/armnn? We don't use Github for the master development branch, just for releases.  There's is a guide to contributing here: https://git.mlplatform.org/ml/armnn.git/about/ContributorGuide.md

Many thanks,
Matthew

**Comment by mablem8** [comment]
Thanks for the feedback @MatthewARM, I will take a look at these links and submit an updated patch.

**Comment by mablem8** [comment]
Hi @MatthewARM,

It actually looks pretty easy to cause an error using the Is..Supported functions. The `reasonIfUnsupported` and `reasonIfUnsupportedMaxLength` parameters have default values of `nullptr` and `1024`, and the only place the Is..Supported functions are called is in the WorkloadFactory where these default values are used. In these cases, the `strncpy` call will never execute due to the `if(truncatedString != nullptr)` guard.

If someone were to write custom code using the Is..Supported functions, it's easy to pass a `reasonIfUnsupported` argument without specifying `reasonIfUnsupportedMaxLength`, in which case the max length will likely be wrong. A simple fix might be to switch the order to `size_t reasonIfUnsupportedMaxLength, char* reasonIfUnsupported` so that `reasonIfUnsupported` can't be specified without also passing a max length.

In any case, the compiler complaint can be fixed more safely than with `memcpy` by replacing the contents of the `if(truncatedString != nullptr)` guard with `std::snprintf(truncatedString, maxLength, fullString);` or anything similar. I pushed the updated code to my fork and will replicate it at mlplatform.org with the proper licensing.

**Comment by MatthewARM** [comment]
Thanks @mablem8 

The IsSupported functions are used pretty extensively in the Android NNAPI wrapper (https://review.mlplatform.org/#/admin/projects/ml/android-nn-driver and https://github.com/ARM-software/android-nn-driver). I take your point about how easy it is to misuse them, but at least it would be a "safe" error - the string wouldn't get copied, it wouldn't be a buffer overflow.

Thanks again for your patches and we'll review them on Gerrit.

Matthew

**Comment by MatthewARM** [comment]
Hi @mablem8 the patch that you have submitted at https://review.mlplatform.org/#/c/ml/armnn/+/773/ doesn't seem to contain any code changes, did something go wrong?

All the best,
Matthew

**Comment by dpackwood** [comment]
Using `aarch64-linux-gnu-g++ (GNU Toolchain for the A-profile Architecture 8.3-2019.03 (arm-rel-8.36)) 8.3.0`
I am getting
~~~~
In function ‘void {anonymous}::CopyErrorMessage(char*, const char*, size_t)’:
error: ‘char* strncpy(char*, const char*, size_t)’ specified bound depends on the length of the source argument [-Werror=stringop-overflow=]
         std::strncpy(truncatedString, fullString, copyLength);
note: length computed here
         size_t copyLength = std::min(maxLength, strlen(fullString));

~~~~

**Comment by MatthewARM** [comment]
Hi @dpackwood, @mablem8 had quite a good fix for that issue, but something went wrong with the Gerrit review and they haven't been back in touch for a while. @mablem8's pull request is here: https://github.com/ARM-software/armnn/pull/141 and it would be great if someone could take over the work of getting it into the master branch of armnn using this gerrit review: https://review.mlplatform.org/#/c/ml/armnn/+/773/ (or just create another review).

All the best,
Matthew

**Comment by ggardet** [comment]
For #195 (build with gcc9), I tried patches from #141 with a patch from #193 (`std::move` problem)
but 6 unit tests fail:
```
[ 1801s] + ./build/UnitTests
[ 1802s] Running 1012 test cases...
[ 1802s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/neon/test/NeonCreateWorkloadTests.cpp(269): [1;31;49merror: in "CreateWorkloadNeon/CreateDepthWiseConvolution2dFloat32NhwcWorkload": check TestNeonTensorHandleInfo(inputHandle, TensorInfo(inputShape, DataType)) has failed[0;39;49m
[ 1802s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/neon/test/NeonCreateWorkloadTests.cpp(270): [1;31;49merror: in "CreateWorkloadNeon/CreateDepthWiseConvolution2dFloat32NhwcWorkload": check TestNeonTensorHandleInfo(outputHandle, TensorInfo(outputShape, DataType)) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(21): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNchwWorkload": check (inputHandle->GetTensorInfo() == inputInfo) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(22): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNchwWorkload": check (outputHandle->GetTensorInfo() == outputInfo) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(21): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNhwcWorkload": check (inputHandle->GetTensorInfo() == inputInfo) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(22): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNhwcWorkload": check (outputHandle->GetTensorInfo() == outputInfo) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(21): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNchwWorkload": check (inputHandle->GetTensorInfo() == inputInfo) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(22): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNchwWorkload": check (outputHandle->GetTensorInfo() == outputInfo) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(21): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNhwcWorkload": check (inputHandle->GetTensorInfo() == inputInfo) has failed[0;39;49m
[ 1808s] /home/abuild/rpmbuild/BUILD/armnn-19.02/src/backends/reference/test/RefCreateWorkloadTests.cpp(22): [1;31;49merror: in "CreateWorkloadRef/CreateConvolution2dFloatNhwcWorkload": check (outputHandle->GetTensorInfo() == outputInfo) has failed[0;39;49m
[ 1813s] 
[ 1813s] [1;31;49m*** 6 failures are detected in the test module "UnitTests"
```

**Comment by MatthewARM** [comment]
Interesting. Can you dig deeper to find out how the tensor infos are different?

Given that the change was to return a tensorinfo by value, instead of by a move,  could indicate a problem with one of the constructors?

**Comment by MikeJKelly** [comment]
Hi,
I think that the unit test issues should have been fixed by https://review.mlplatform.org/c/ml/armnn/+/1333

**Comment by psyhtest** [comment]
I've just hit the same issue compiling v19.05 with GCC 8.3.0 on Raspberry Pi 4:
```
/home/anton/CK_TOOLS/lib-armnn-19.05-gcc-8.3.0-neon-rel.19.05-tflite-linux-32/src/src/armnn/LayerSupport.cpp: In function ‘void {anonymous}::CopyErrorMessage(char*, const char*, size_t)’:                 
/home/anton/CK_TOOLS/lib-armnn-19.05-gcc-8.3.0-neon-rel.19.05-tflite-linux-32/src/src/armnn/LayerSupport.cpp:28:21: error: ‘char* strncpy(char*, const char*, size_t)’ specified bound depends on the length
 of the source argument [-Werror=stringop-overflow=]                                                                                                                                                        
         std::strncpy(truncatedString, fullString, copyLength);                                                                                                                                             
         ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                              
/home/anton/CK_TOOLS/lib-armnn-19.05-gcc-8.3.0-neon-rel.19.05-tflite-linux-32/src/src/armnn/LayerSupport.cpp:27:55: note: length computed here
         size_t copyLength = std::min(maxLength, strlen(fullString));
                                                 ~~~~~~^~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

**Comment by ggardet** [comment]
It seems to be fixed in 19.08.

**Comment by MatthewARM** [comment]
Thanks @ggardet. Closing.
